sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/TemporalJoin.java

```diff
@@ -0,0 +1,491 @@
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.state.OrderedListState;
+import org.apache.beam.sdk.state.StateSpec;
+import org.apache.beam.sdk.state.StateSpecs;
+import org.apache.beam.sdk.state.TimeDomain;
+import org.apache.beam.sdk.state.TimerMap;
+import org.apache.beam.sdk.state.TimerSpec;
+import org.apache.beam.sdk.state.TimerSpecs;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindow;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.apache.beam.sdk.values.TimestampedValue.TimestampedValueCoder;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.PeekingIterator;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * This class allows joining two {@link PCollection}s using temporal conditions.
+ *
+ * <p>Several types of joins are supported:
+ *
+ * <p><b>As Of Joins</b> Join each element of the main PCollection with the element/s of the
+ * right-hand PCollection having the closest timestamp before the main element's timestamp.
+ *
+ * <p>For example, consider two PCollections representing currency trades and quotes. For each trade
+ * we want to find the most-recent quote for that currency that has a timestamp earlier than the
+ * trade; this will give us the currency price at the time of the trade.
+ *
+ * <pre>{@code
+ * PCollection<KV<Currency, Double>> quotes = readQuotes();
+ * PCollection<KV<Currency, Trade>> trades = readTrades();
+ * PCollection<Trade> tradesWithPrices =
+ *   trades.apply(TemporalJoin.innerJoin(quotes))
+ *         .apply(Values.create())
+ *         .apply(MapElements.to(trades()).via(result -> result.getMainElement().getValue().withPrice(result.getAgainstElements().get(0))));
+ * }</pre>
+ *
+ * <b>Within Joins</b> Join each element of the main PCollection with all elements of the right-hand
+ * PCollection that happened within a specified duration before the main element's timestamp. For
+ * example:
+ *
+ * <pre>{@code
+ * trades.apply(TemporalJoin.innerJoin(quotes).within(Duration.standardHours(1)));
+ * }</pre>
+ *
+ * Will return all quotes that happened within the hour before each trade.
+ *
+ * <p>The PCollection's built-in timestamp dimension is used for all of these joins. The user can
```

Comment:
maybe add a little more , something like "...joins. If you have not specifically specified this value using withoutputtimestamp, then this value is derived  from the Source....